### PR TITLE
#872 Generate Data Filter columns

### PR DESCRIPTION
**Pull request by CarlosNZ**
Implements #872, pretty much exactly as described there.

Script is working well, and I've added an endpoint so we can run it manually:

`POST: /admin/generate-filter-data-fields?table=product` (if `table` parameter is omitted, it'll run on ALL tables with filter data columns.

See snapshot `/conforma-templates/dev/snapshots/AA_DataViewAutoGenFields.zip` to see example definitions and test it out.

Added an option to `modifyRecord` so we can run a filter data update whenever a record is added to the system.

Once we're happy with this, we'll just need to add some configurations to Angola and they'll have filter-able everything!

---

**Update**: Have added an additional "extractNumber" function to "evaluatorFunctions" which allows us to do things like pull the numerical value out of fields like "shelf life" (stored as "3 Meses" for example), which means we can filter by the number range. (Utilises the number filter in https://github.com/openmsupply/conforma-web-app/pull/1364 -- load that branch in front-end to see it in action)

### CODE REVIEW

**Comment by nmadruga** [comment]
Hmmm just so I can understand this one better - is this used to update Filters that should be available for all DataViews in the snapshot? Would this work to update the latest snapshot to have the filters working or something else is required?

I'm planning to Review tomorrow when we have the other base PRs merged in.

**Comment by CarlosNZ** [comment]
> Hmmm just so I can understand this one better - is this used to update Filters that should be available for all DataViews in the snapshot? 

Did you read the explanation of the problem that this solves? (#872)

> Would this work to update the latest snapshot to have the filters working or something else is required?

It allows us to maintain and modify whatever filters we want without having to worry about every possible filter in advance and define it precisely in modifyRecord. Basically, we should be able to filter anything (anything extractable from an evaluator query, anyway), but we don't need to worry about it at all in the template configuration, which would be a massive headache every time anything changed slightly.

I can go through it with you in person if you'd like further clarification or you'd like to offer suggestions.

**Review by nmadruga — COMMENTED** [review]
Thanks, Carl. I had a proper look on the issue #872 to understand better the idea here. Although I'm still finding it difficult to Review. I think I just have too much in my head now so will come back to this Tomorrow and do a proper testing and finish reading. Pretty much stopped at reading and comprehending these 2:
src/components/data_display/generateFilterDataFields/databaseMethods.ts
src/components/data_display/generateFilterDataFields/generateFilterDataFields.ts

So forgive me this will be pending on my list still. Maybe @fergie-nz could also have a look and if there is time tomorrow go through with you on a open session. I think it would be helpful indeed!

**Comment by CarlosNZ** [comment]
> So forgive me this will be pending on my list still

It really only became apparent to me when I was doing the other filtering PRs how much of a pain in the ass it was going to be to maintain these additional "text" versions of existing data columns. So I suspect you haven't yet experienced the pain of that realisation, and hence the need for a solution. 😜

Probably best if I spend 5 minutes with you tomorrow clarifying your confusion and then go from there?

File: src/components/data_display/helpers.ts [hunk 1]
```diff
@@ -198,6 +199,8 @@ const buildColumnList = (
       ? 'tableViewExcludeColumns'
       : type === 'DETAIL'
       ? 'detailViewExcludeColumns'
+      : dataView.filterIncludeColumns === null && dataView.filterExcludeColumns === null
+      ? 'tableViewExcludeColumns'
```
Comment (CarlosNZ) [inline src/components/data_display/helpers.ts:203]: This a small bug fix for earlier PRs, unrelated to this functionality.

**Review by nmadruga — CHANGES_REQUESTED** [review]
I can see that most of the work is defined in `src/components/data_display/generateFilterDataFields/generateFilterDataFields.ts` but is still a little no clear. And I imagine for someone else looking at this for the first time. Would you be able to include a Overview description of this work please. What I suggest is create a new document on Front-end to explain how this helps to generate the "default"(?) filters

Some small things I noticed while testing instance https://conforma-demo.msupply.org:50000:
- The first one observed has stopped happening was a tiny thing with dates filter XD
- The loading after you click one filter option OR when sorting a column is showing weirdly. The list is visible on the bottom and the loading takes most of the space

File: src/components/data_display/generateFilterDataFields/databaseMethods.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    }
+  },
+  addOrUpdateColumn: async (table: string, columnName: string, dataType: string) => {
+    const text = `
+      ALTER TABLE "${table}"
+        DROP COLUMN IF EXISTS ${columnName},
+        ADD COLUMN ${columnName} ${dataType};
+    `
+    try {
+      await DBConnect.query({ text })
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/data_display/generateFilterDataFields/databaseMethods.ts:81]: Cos you can't use them for schema-related things like table names and column names. :(

**Comment by CarlosNZ** [comment]
> I can see that most of the work is defined in `src/components/data_display/generateFilterDataFields/generateFilterDataFields.ts` but is still a little no clear. And I imagine for someone else looking at this for the first time. Would you be able to include a Overview description of this work please. What I suggest is create a new document on Front-end to explain how this helps to generate the "default"(?) filters

From https://github.com/openmsupply/conforma-web-app/pull/1364#issue-1393382578: 

Screen Shot 2022-10-11 at 3 15 24 PM

Planning to do on the feature branch before we merge to develop. But want everything to be settled first, so I don't have to re-write it due to changes.

**Comment by nmadruga** [comment]
> > I can see that most of the work is defined in `src/components/data_display/generateFilterDataFields/generateFilterDataFields.ts` but is still a little no clear. And I imagine for someone else looking at this for the first time. Would you be able to include a Overview description of this work please. What I suggest is create a new document on Front-end to explain how this helps to generate the "default"(?) filters
> 
> From [openmsupply/conforma-web-app#1364 (comment)](https://github.com/openmsupply/conforma-web-app/pull/1364#issue-1393382578):
> 
> Screen Shot 2022-10-11 at 3 15 24 PM
> 
> Planning to do on the feature branch before we merge to develop. But want everything to be settled first, so I don't have to re-write it due to changes.

That sounds lovely. Whenever is a good time for you

**Review by nmadruga — APPROVED** [review]

**Pull request merged**